FastSimulation/FastSimProducer/plugins/FastSimProducer.cc

```diff
@@ -0,0 +1,224 @@
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/LuminosityBlock.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+// data formats
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+#include "SimDataFormats/TrackingHit/interface/PSimHitContainer.h"
+#include "SimDataFormats/Track/interface/SimTrackContainer.h"
+#include "SimDataFormats/Vertex/interface/SimVertexContainer.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/Math/interface/LorentzVector.h"
+
+// fastsim
+#include "FastSimulation/Utilities/interface/RandomEngineAndDistribution.h"
+#include "FastSimulation/Geometry/interface/Geometry.h"
+#include "FastSimulation/Layer/interface/Layer.h"
+#include "FastSimulation/Decayer/interface/Decayer.h"
+#include "FastSimulation/Propagation/interface/LayerNavigator.h"
+#include "FastSimulation/NewParticle/interface/Particle.h"
+#include "FastSimulation/FastSimProducer/interface/ParticleFilter.h"
+#include "FastSimulation/Particle/interface/ParticleTable.h"  // TODO: get rid of this
+#include "FastSimulation/InteractionModel/interface/InteractionModel.h"
+#include "FastSimulation/InteractionModel/interface/InteractionModelFactory.h"
+#include "FastSimulation/FastSimProducer/interface/ParticleLooper.h"
+
+// other
+
+class FastSimProducer : public edm::stream::EDProducer<> {
+public:
+
+    explicit FastSimProducer(const edm::ParameterSet&);
+    ~FastSimProducer(){;}
+
+private:
+
+    virtual void produce(edm::Event&, const edm::EventSetup&) override;
+
+    edm::EDGetTokenT<edm::HepMCProduct> genParticlesToken_;
+    fastsim::Geometry geometry_;
+    double beamPipeRadius_;
+    fastsim::ParticleFilter particleFilter_;
+    fastsim::Decayer decayer_;
+    std::vector<std::unique_ptr<fastsim::InteractionModel> > interactionModels_;
+    std::map<std::string,fastsim::InteractionModel *> interactionModelMap_;
+    edm::IOVSyncValue iovSyncValue_;
+    static const std::string MESSAGECATEGORY;
+};
+
+const std::string FastSimProducer::MESSAGECATEGORY = "FastSimulation";
+
+FastSimProducer::FastSimProducer(const edm::ParameterSet& iConfig)
+    : genParticlesToken_(consumes<edm::HepMCProduct>(iConfig.getParameter<edm::InputTag>("src"))) 
+    , geometry_(iConfig.getParameter<edm::ParameterSet>("detectorDefinition"))
+    , beamPipeRadius_(iConfig.getParameter<double>("beamPipeRadius"))
+    , particleFilter_(iConfig.getParameter<edm::ParameterSet>("particleFilter"))
+{
+
+    //----------------
+    // define interaction models
+    //---------------
+    const edm::ParameterSet & modelCfgs = iConfig.getParameter<edm::ParameterSet>("interactionModels");
+    for( const std::string & modelName : modelCfgs.getParameterNames())
```

Comment:
Thank you for the comment. So far I am not aware of any cases where this might be a problem. So far, not all the plugins (interaction models) are defined yet, so I will keep it in mind.